starknet/test/safeUtils.ts

```diff
@@ -0,0 +1,372 @@
+  to: string;
+  value: string | number | BigNumber;
+  data: string;
+  operation: number;
+}
+
+export interface SafeTransaction extends MetaTransaction {
+  safeTxGas: string | number;
+  baseGas: string | number;
+  gasPrice: string | number;
+  gasToken: string;
+  refundReceiver: string;
+  nonce: string | number;
+}
+
+export interface SafeSignature {
+  signer: string;
+  data: string;
+}
+
+const encodeMetaTransaction = (tx: MetaTransaction): string => {
+  const data = utils.arrayify(tx.data);
+  const encoded = utils.solidityPack(
+    ['uint8', 'address', 'uint256', 'uint256', 'bytes'],
+    [tx.operation, tx.to, tx.value, data.length, data],
+  );
+  return encoded.slice(2);
+};
+
+export const encodeMultiSend = (txs: MetaTransaction[]): string => {
+  return '0x' + txs.map((tx) => encodeMetaTransaction(tx)).join('');
+};
+
+export const buildMultiSendSafeTx = (
+  multiSend: Contract,
+  txs: MetaTransaction[],
+  nonce: number,
+  overrides?: Partial<SafeTransaction>,
+): SafeTransaction => {
+  return buildContractCall(multiSend, 'multiSend', [encodeMultiSend(txs)], nonce, true, overrides);
+};
+
+export const calculateSafeDomainSeparator = (safe: Contract, chainId: BigNumberish): string => {
+  return utils._TypedDataEncoder.hashDomain({ verifyingContract: safe.address, chainId });
+};
+
+export const preimageSafeTransactionHash = (
+  safe: Contract,
+  safeTx: SafeTransaction,
+  chainId: BigNumberish,
+): string => {
+  return utils._TypedDataEncoder.encode(
+    { verifyingContract: safe.address, chainId },
+    EIP712_SAFE_TX_TYPE,
+    safeTx,
+  );
+};
+
+export const calculateSafeTransactionHash = (
+  safe: Contract,
+  safeTx: SafeTransaction,
+  chainId: BigNumberish,
+): string => {
+  return utils._TypedDataEncoder.hash(
+    { verifyingContract: safe.address, chainId },
+    EIP712_SAFE_TX_TYPE,
+    safeTx,
+  );
+};
+
+export const calculateSafeMessageHash = (
+  safe: Contract,
+  message: string,
+  chainId: BigNumberish,
+): string => {
+  return utils._TypedDataEncoder.hash(
+    { verifyingContract: safe.address, chainId },
+    EIP712_SAFE_MESSAGE_TYPE,
+    {
+      message,
+    },
+  );
+};
+
+export const safeApproveHash = async (
+  signer: Signer,
+  safe: Contract,
+  safeTx: SafeTransaction,
+  skipOnChainApproval?: boolean,
+): Promise<SafeSignature> => {
+  if (!skipOnChainApproval) {
+    if (!signer.provider) throw Error('Provider required for on-chain approval');
+    const chainId = (await signer.provider.getNetwork()).chainId;
+    const typedDataHash = utils.arrayify(calculateSafeTransactionHash(safe, safeTx, chainId));
+    const signerSafe = safe.connect(signer);
+    await signerSafe.approveHash(typedDataHash);
+  }
+  const signerAddress = await signer.getAddress();
+  return {
+    signer: signerAddress,
+    data:
+      '0x000000000000000000000000' +
+      signerAddress.slice(2) +
+      '0000000000000000000000000000000000000000000000000000000000000000' +
+      '01',
```

Comment:
lol, that's odd :p some Safe-specific data?

Reply:
yep, `safeUtils.ts` was taken from their repo so not sure exactly whats going on there

Reply:
what would be nice is to easily be able to identify which files were copied / which ones were copied but edited / which ones are new

Reply:
we shouldn't need to audit/ review anything in there as we can assume the safe contracts work as expected

Reply:
exactly, I mean they're in the `test` folder so I guess auditors won't look at them anyway, but what I mean is having a file prefix or a folder prefix for "code that is not ours" would be nice imho

Reply:
ill create an external sub dir